### DISCUSSION TITLE
Support profile pictures for users

### DISCUSSION THREAD

**Discussion by kanpov** [keycloak/keycloak#30428] [Ideas]
There's this new Account Console v3 in Keycloak 24.0 and yet it only supports text attributes. I've yet to find a halfway decent solution of how to allow users to upload profile pictures to their Keycloak account, even though this seems like a really obvious feature to have.

I've found:
1. A vague reference to how this is "part of some plan" in an old discussion, linking to an issue on a repo that doesn't exist
2. A [proof of concept extension](https://github.com/thomasdarimont/keycloak-avatar-minio-extension) for this that hasn't been updated in more than 5 years

Strange to see such an advanced identity platform lacking this.

**Comment by ssilvert** [comment]
@kanpov We do support the OIDC picture claim in both account and admin console.

The simplest way to see this working is to add a user profile attribute called `picture`.  Then a user or administrator can add a URL to the `picture` field and you will see the user's profile pic in the UI.

We have decided not to implement a way to edit or upload a profile picture at this time.  That would add some complexity to Keycloak that we don't want to take on right now.

**Reply by helmut72** [reply]
> We have decided not to implement a way to edit or upload a profile picture at this time.

This means a user can change most of his data directly in Keycloak and only for the picture another preference page somewhere else is needed?

**Comment by carlchen2007** [comment]
Any new updates? Personally, I think a way to upload a photo would be quite efficient as not all users have an uploaded profile picture somewhere or if so, it is too complex for the simple user to find the profile picture URL. And even considering that it is a huge security risk if I hand the URL to other clients as this would provide attack surface since any link could be added. Did I miss anything?

**Reply by coopbri** [reply]
Agreed, UX is poor if users have to find an image URL and plug it in manually

**Reply by seboettg** [reply]
There are also other challenges in this context: How do I ensure that the photos have a uniform size with an external URL?

**Reply by ssilvert** [reply]
@seboettg The [PatternFly avatar tag](https://www.patternfly.org/components/avatar/) takes care of scaling.

**Comment by stianst** [comment]
Specifying a profile picture from a URL should only allow trusted sources of images. That can be done by adding a validator on the attribute, as well as CSP headers etc. for any applications that is displaying an image.

This for sure it's not a great way to do it, and ideally Keycloak should provide some capabilities of uploading profile pictures.

**Comment by coopbri** [comment]
I found https://github.com/Yreus6/keycloak-avatar-minio-extension (a slightly more recent fork of the proof of concept mentioned in the OP), however couldn't get this to work in Keycloak v26. It would be really cool if the foundation of this plugin was part of Keycloak core

**Comment by ahmadofski** [comment]
Is there a reasonably painless way of implementing this? We're looking to use Keycloak on our webapp and are required to have a profile picture attribute

**Reply by n1cc00** [reply]
Wasn't able to find another solution other than this (URL):
https://www.keycloak.org/ui-customization/avatars

Or develop a plugin for it..

**Reply by ahmadofski** [reply]
> Wasn't able to find another solution other than this (URL):
> https://www.keycloak.org/ui-customization/avatars
> 
> Or develop a plugin for it..

Not ideal TBH
